hybrasyl/World.cs

```diff
@@ -2376,11 +2474,371 @@ private void PacketHandler_0x30_MoveUIElement(Object obj, ClientPacket packet)
         private void PacketHandler_0x3B_AccessMessages(Object obj, ClientPacket packet)
         {
             var user = (User) obj;
-            var messagePacket = new ServerPacket(0x31);
-            messagePacket.WriteByte(0x01);
-            messagePacket.WriteUInt16(0x00);
-            messagePacket.WriteByte(0x00);
-            user.Enqueue(messagePacket);
+            var response = new ServerPacket(0x31);
+            var action = packet.ReadByte();
+
+            switch (action)
+            {
+                case 0x01:
+                {
+                    // Display board list
+                    response.WriteByte(0x01);
+
+                    // TODO: This has the potential to be a somewhat expensive operation, optimize this.
+                    var boardList =
+                        Messageboards.Values.Where(mb => mb.CheckAccessLevel(user.Name, BoardAccessLevel.Read));
+
+                    // Mail is always the first board and has a fixed id of 0
+                    response.WriteUInt16((ushort) (boardList.Count() + 1));
+                    response.WriteUInt16(0);
+                    response.WriteString8("Mail");
+                    foreach (var board in boardList)
+                    {
+                        response.WriteUInt16((ushort) board.Id);
+                        response.WriteString8(board.Name);
+                    }
+                    response.TransmitDelay = 600; // This is so the 'w' key in the client works
+                    // Without this, the messaging panel is a jittery piece of crap that never opens
+                }
+                    break;
+                case 0x02:
+                {
+                    // Get message list
+                    var boardId = packet.ReadUInt16();
+                    var startPostId = packet.ReadInt16();
+
+                    if (boardId == 0)
+                    {
+                        user.Enqueue(user.Mailbox.RenderToPacket());
+                        return;
+                    }
+                    else
+                    {
+                        Board board;
+                        if (MessageboardIndex.TryGetValue(boardId, out board))
+                        {
+                            user.Enqueue(board.RenderToPacket());
+                            return;
+                        }
+                        else
+                        {
+                            return;
+                        }
+                    }
+                }
+                case 0x03:
+                {
+                    // Get message
+                    var boardId = packet.ReadUInt16();
+                    var postId = packet.ReadInt16();
+                    var messageId = postId - 1;
+                    var offset = packet.ReadSByte();
+                    Message message = null;
+                    var error = string.Empty;
+                    if (boardId == 0)

```

Comment:
I don't like this giant if/else with a case statement nested within a case statement. some refactoring should happen (not critical).